src/sulogin.c

```diff
@@ -160,25 +158,25 @@ main(int argc, char **argv)
 		 * it will work with standard getpass() (no NULL on EOF).
 		 * --marekm
 		 */
-		if ((NULL == cp) || ('\0' == *cp)) {
-			erase_pass (cp);
+		if ((NULL == pass) || ('\0' == *pass)) {
+			erase_pass (pass);
 			(void) puts ("");
 #ifdef	TELINIT
 			execl (PATH_TELINIT, "telinit", RUNLEVEL, (char *) NULL);
 #endif
 			exit (0);
 		}
-		STRTCPY(pass, cp);
-		erase_pass (cp);
 
-		if (valid (pass, &pwent)) {	/* check encrypted passwords ... */
-			break;	/* ... encrypted passwords matched */
+		/* check encrypted passwords ... */
+		done = valid(pass, &pwent);
+		if (!done) {
+			/* ... encrypted passwords did not match */
+			sleep (2);
+			(void) puts (_("Login incorrect"));
 		}
 
-		sleep (2);
-		(void) puts (_("Login incorrect"));
+		erase_pass (pass);
```

Comment:
This would be more appropriate before the conditional from above.  Otherwise we have the password in memory for 2 seconds unnecessarily.

Reply:
Good point, thanks! Applied.